Initial Parsing of BMV2 JSON programs with protobuf

KinanBab opened the pull request:
**Work in progress**

**tl;dr:** _parsing will be done in two phases due to genericity of constructs, protobuf definitions for the first phase are implemented here, second phase and connecting c++ glue are TODO. tests fail because protobuf replaces null / missing fields with default values._

This pull request includes the following components:
1. protobuf definitions specifying the overall structure of the bmv2 json format
2. a simple c++ program that parses and dump json program using protobuf
3. testing protobuf definition via the c++ program and subset diff against input json

The protobuf definitions only cover the overall structure of the json format. A lot is still missing:
- some pieces of the json format contains information that we do not need for analysis
- other pieces are important for analysis but are too generic (e.g. an expression in the JSON format has different attributes and types depending on the concrete type of expression), which is hard to express in protobuf

For the second type of missing information, the plan is to define concrete and detailed protobuf definitions for each concrete type. These detailed definitions are separate from the overall structure definitions. The overall structure uses placeholder unconstrained protobuf constructs (e.g Struct, Value, ValueList, Any) in locations that are generic. The c++ analysis and traversal code later re-parses the unconstrained placeholder data types into the detailed ones based on their dynamic structure and type. This is not yet implemented.

The testing pipeline is available under /testing. Currently, the tests fail, because protobuf fills in null and missing values with default values per defined type (a null in lieu of a string is parsed as ""), causing the subset diff to fail when comparing the "" with the respective null. This will be fixed appropriately after some thinking and experimentation.

To compile the c++ main file
```
bazel build :main
```

To run the tests
```
bazel test --test_output=all testing:parsing
```

KinanBab committed:
Add small protobuf test file for parsing JSON using make

KinanBab committed:
Add initial bazel files and hello world main

KinanBab committed:
Use Protobuf through Bazel

* create a bazel package for protobuf definitions
* bazel package specifies the protobuf compilation via protoc
* main bazel package requires protofbuf generated source

KinanBab committed:
Formatting and clean up

KinanBab committed:
Use c++11

KinanBab committed:
Add protobuf definitions for simplified bmv2 JSON format

KinanBab committed:
Add simple C++ stub for testing protobuf definitions

KinanBab committed:
Add subset diff testing for protobuf parsing through bazel

stefanheule commented on src/protobuf/bmv2/table.proto:
```diff
@@ -0,0 +1,43 @@
+  int32 id = 2;
+  SourceInfo source_info = 3;
+  repeated TableKey key = 4;
+  string match_type = 5; // make into an enum
```
Should this be a TODO? At Google we generally start todo's with "TODO(username): ...".

stefanheule commented on src/protobuf/bmv2/table.proto:
```diff
@@ -0,0 +1,43 @@
+  SourceInfo source_info = 3;
+  repeated TableKey key = 4;
+  string match_type = 5; // make into an enum
+  string type = 6; // simple
```
What does simple mean here?

stefanheule commented on src/protobuf/bmv2/table.proto:
```diff
@@ -0,0 +1,43 @@
+  string match_type = 5; // make into an enum
+  string type = 6; // simple
+  int32 max_size = 7;
+  repeated int32 action_ids = 8;
```
What's the difference beteen this and actions?

stefanheule commented on src/protobuf/bmv2/table.proto:
```diff
@@ -0,0 +1,43 @@
+}
+
+message TableKey {
+  string match_type = 1;
```
Should this be an enum?

stefanheule commented on src/protobuf/bmv2/table.proto:
```diff
@@ -0,0 +1,43 @@
+  string match_type = 1;
+  string name = 2;
+  string mask = 3;
+  repeated string target = 4;
```
What is target and mask?

stefanheule replied:
I guess generally we might need some comments explaining some of the less obvious fields.

stefanheule commented on testing/BUILD:
```diff
@@ -0,0 +1,21 @@
+    ]
+)
+
+subset_diff_test(
```
One big advantage of simple diff testing is that you get to see the output, this is lost with this definition of subset_diff_test. That is, I can't see in this review what the protobuf looks like for the given P4 program, which seems like a big downside.

Could we make it so that the protobuf is the expected output and is what is checked in?

KinanBab replied:
I am going to check in the output of the protobuf parsing then dumping, we can use this for golden files testing as well as the subset diff testing

stefanheule commented on testing/BUILD:
```diff
@@ -0,0 +1,21 @@
+)
+
+subset_diff_test(
+    name = "table_test",
```
Another problem is that this is not going from a P4 program all the way to a protobuf, but I think that is what we would want, is it not?

Here's what I'm envisioning:

```
diff_test(
    name = "reflector_test",
    actual = ":reflector",
    expected = "testdata/reflector.expected",
)

run_parser(
    name = "reflector",
    src = "testdata/reflector.p4",
)
```

There is a rule `run_parser` that takes a P4 program as input and runs the parser, and outputs the protobuf. Then, based on that, we can have a diff-test that compares against the expected file.

We can still do subset diffing if we want as part of the run_parser rule, which would output some error in that case instead of (or maybe in addition to) the protobuf. But if subset diffing works, then the output should be the protobuf.

What do you think?

KinanBab replied:
Yes, sounds like a good idea, I will find a way to implement it

stefanheule reviewed (commented):
This is looking great, thanks for sending it out!  I definitely would encourage us to try and submit something small, that does not yet have all the features, and then expand on that. We don't need to support things like the parser or deparser, for instance, we are likely not going to look at those for a while.

stefanheule commented:
> ## default_entry
> https://github.com/googleinterns/p4-symbolic-test-gen/blob/b0cbcbef89a2cd26d61b8eefb1cc6f9428487ea4/src/protobuf/bmv2/table.proto#L35-L40
> 
> ###### This comment was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in [b0cbcbe](https://github.com/googleinterns/p4-symbolic-test-gen/commit/b0cbcbef89a2cd26d61b8eefb1cc6f9428487ea4) in #3. cc @googleinterns.

What is this bot doing? Did we enable this?

KinanBab committed:
Add more documentation for bmv2 protobuf definitions

KinanBab committed:
move samples into testing directory

KinanBab committed:
Improve bazel testing of protobuf parsing

* create bazel custom rules for running p4c on p4 programs and getting bmv2 json files
* test rules depend on p4c rules
* test rules use main.cc to parse and dump bmv2 json file, then check that they are a subset of the input json file

KinanBab committed:
Check in p4c bmv2 sample json file and the protobuf dump of it

KinanBab committed:
Finalize subset diff tests

* subset diff equates a null / missing field in the p4c output with default values in protobuf output
* check in all protobuf and p4c json files under testing/output
* fixes in protobuf defintions

KinanBab committed:
Lint and add license

smolkaj commented:
This looks great; sorry I'm a bit late to the party.

First some general remarks.
* You may find [this script](https://github.com/p4lang/p4-constraints/blob/master/format.sh) useful, which will format your code according to Google's style guide. For the last line in the script to work, you need to adjust your BUILD and WORKSPACE files.
* Now that we have some tests, we may want to run them as part of CI. You can use the Github Actions from [here](https://github.com/p4lang/p4-constraints/tree/master/.github/workflows) if you'd like. (It may come in handy to know that `bazel build //...` will build all targets in your project recursively, and similar for `bazel test`.

smolkaj commented on BUILD:
```diff
@@ -0,0 +1,32 @@
+    visibility = ["//test:__pkg__"]
+)
+
+test_suite(
```
There is nothing wrong with this, but it may be handy to know that you can always run all tests in a project using `bazel test //...`, so if that's all you want to do there is no need to hardcode things.

smolkaj commented on p4c.bzl:
```diff
@@ -0,0 +1,48 @@
+# limitations under the License.
+
+def _run_p4c_impl(ctx):
+    fname = ctx.attr.name + "-bazel-tmp-output" + "/" + \
```
Just a nit, but the more idiomatic way to do this would be to write
```
"".join([x, y, z])
```

smolkaj commented on p4c.bzl:
```diff
@@ -0,0 +1,48 @@
+    ctx.actions.run_shell(
+        inputs = [ctx.file.p4program],
+        outputs = [json_output_file],
+        use_default_shell_env = True,
```
Interesting. Did you add this so p4c finds `cc`? Just curious

KinanBab replied:
p4c needs $PATH variable for several things, one particular thing is the c++ preprocesser. Without $PATH the p4c python entry code throws errors because it tries to access env['PATH'] which is not set.

smolkaj replied:
Got it. Nice.

smolkaj commented on p4c.bzl:
```diff
@@ -0,0 +1,48 @@
+        "p4program": attr.label(
+            doc = "Input .p4 program to compile into a bmv2 JSON",
+            mandatory = True,
+            allow_single_file = True,
```
P4 programs often `#include` other p4 programs, so you will likely have to add another argument (maybe `deps`) for those files later. No need to do so right now, just something to have in mind.

KinanBab replied:
good idea

KinanBab replied:
I opened an issue for this for the future, will revisit this later down the road.

smolkaj commented on src/main.cc:
```diff
@@ -0,0 +1,57 @@
+#include <string>
+
+// #include "absl/strings/str_cat.h"
+// #include "absl/strings/string_view.h"
```
Delete?

smolkaj commented on src/main.cc:
```diff
@@ -0,0 +1,57 @@
+  GOOGLE_PROTOBUF_VERIFY_VERSION;
+
+  std::string* input = new std::string();
+  ReadStdin(input);
```
See above; prefer `std::string input = ReadStdin()`.

smolkaj commented on src/main.cc:
```diff
@@ -0,0 +1,57 @@
+
+#include "src/protobuf/bmv2/program.pb.h"
+
+void ReadStdin(std::string* str) {
```
Prefer using a return value: http://go/cstyle#Inputs_and_Outputs

smolkaj commented on src/main.cc:
```diff
@@ -0,0 +1,57 @@
+}
+
+int main() {
+  // verify link and compile versions are the same
```
Nit: the google style for comments is `// Upper case terminated by period.`

Fix here and throughout.

smolkaj commented on p4c.bzl:
```diff
@@ -0,0 +1,48 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
Please add a one line summary what the file is good for.

smolkaj commented on src/main.cc:
```diff
@@ -0,0 +1,57 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
Please add a one line summary what the file is doing. Here and for source files in general. If you have both a header file and a `cc` file, it's enough to put the comment in the header file.

smolkaj commented on src/protobuf/bmv2/header.proto:
```diff
@@ -0,0 +1,46 @@
+
+import "google/protobuf/struct.proto";
+
+// A header type definition
```
Nit: add '.' (here and elsewhere).

smolkaj replied:
A P4 header type definition?

smolkaj commented on src/protobuf/bmv2/header.proto:
```diff
@@ -0,0 +1,46 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
Add one line summary.

smolkaj commented on src/protobuf/bmv2/header.proto:
```diff
@@ -0,0 +1,46 @@
+// limitations under the License.
+
+syntax = "proto3";
+
```
We should decide on a namespace for this project and add it here.
In general, everything needs to go in a namespace.

smolkaj commented on src/protobuf/bmv2/parser.proto:
```diff
@@ -0,0 +1,46 @@
+// limitations under the License.
+
+syntax = "proto3";
+
```
As above, let's put this in a namespace: `package = ...`. Here and in all proto files.

smolkaj commented on src/protobuf/bmv2/primitives.proto:
```diff
@@ -0,0 +1,41 @@
+
+// A value/expression is a struct
+// it contains a type: field, hexstr, regular, runtime_data, header, expression, op, assign
+// and a value: array, string, string, int, string, (nested expression), (struct with op, left, and right), (no value only parameters)
```
Lets keep things <= 80 columns. No need to do this by hand though, it's better to use clang-format for that.

smolkaj commented on src/protobuf/bmv2/program.proto:
```diff
@@ -0,0 +1,58 @@
+// constrained protobuf definitions
+// For example, expressions in P4 look very different according
+// to their type, they may have a different set of fields
+// and may assign different type to the same field
```
Nit: use periods.

smolkaj commented on src/protobuf/bmv2/table.proto:
```diff
@@ -0,0 +1,107 @@
+// Copyright 2020 Google LLC
```
Are we ever gonna use any of these files individually? If not, have you considered combining them into a single file?
If these files all serve a single purpose, having them in a single file with a comment at the top explaining the file's purpose may make it easier to digest for a reader.

As a general principle, we optimize for code readability over ease of writing the code.

stefanheule replied:
I already suggested this and there is a bug to do this. It'll be done in a separate CL :)

smolkaj commented on test/BUILD:
```diff
@@ -0,0 +1,69 @@
+)
+
+# rules for compiling p4 programs to bmv2 json files
+run_p4c(
```
What's the output of this rule? Ideally, I should be able to tell what the inputs and outputs are without looking at the definition of `run_p4c`. 
Maybe make the output a parameter of the rule. Something like `p4info = ` or `bmv2_json = ...`.

smolkaj commented on test/BUILD:
```diff
@@ -0,0 +1,69 @@
+
+# rules for compiling p4 programs to bmv2 json files
+run_p4c(
+  name = "p4c_table",
```
Nit: the typical convention is to use suffixes instead of prefixes, e.g. "table_p4c".
We also typically use names that capture what output is generated.

smolkaj commented on test/bmv2_parsing_check.bzl:
```diff
@@ -0,0 +1,67 @@
+    )
+
+subset_diff_test = rule(
+    doc = """Computes a smart subset diff between two JSON objects.
```
Nit: the convention is to insert a newline after this.

stefanheule commented on src/protobuf/bmv2/primitives.proto:
```diff
@@ -0,0 +1,41 @@
+// and a value: array, string, string, int, string, (nested expression), (struct with op, left, and right), (no value only parameters)
+
+// Describes the definition of a variable or action parameter
+message RuntimeData {
```
Would VariableDefinition be a better name for this?

stefanheule commented on src/protobuf/bmv2/primitives.proto:
```diff
@@ -0,0 +1,41 @@
+
+// Contains source code information about the location
+// in the source where the relevant construct was defined
+message SourceInfo {
```
Maybe SourceLocation?

stefanheule commented on src/protobuf/bmv2/program.proto:
```diff
@@ -0,0 +1,58 @@
+  // The packet overall header, usually defined
+  // as a collection of the above header type
+  repeated Header headers = 2;
+  // All parsers defined in the program
```
There can be more than one parser? That seems surprising.

stefanheule commented on src/protobuf/bmv2/program.proto:
```diff
@@ -0,0 +1,58 @@
+import "src/protobuf/bmv2/table.proto";
+
+
+// Specifies the overal structure of a p4 program json
```
Lets follow the style guide and use full sentences that end with a full stop "." (see https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar). Might seem like a nitpick, but it's good to get into the habit of writing clear and easy to read comments :)

Please fix here and everywhere.

smolkaj commented on test/bmv2_parsing_check.bzl:
```diff
@@ -0,0 +1,67 @@
+        "input": attr.label(
+            doc = "The input JSON file, or a rule producing it (typically invoking p4c)",
+            mandatory = True,
+            allow_single_file = True
```
Nit: our formatting tools would insert a trailing ",", here and everywhere where you use lists. This is makes it easier to modify things in the future.

stefanheule replied:
Probably fine to ignore for now and instead (in a separate PR) enable auto-formatting.

KinanBab replied:
Yes, I have an issue open for this and would prefer to do it in the next PR.

stefanheule commented on src/protobuf/bmv2/table.proto:
```diff
@@ -0,0 +1,107 @@
+// Copyright 2020 Google LLC
```
This is an overall comment, but I think it would be worth keeping all protos about representing bmv2 in a single file. It makes it much easier to read. What do you think?

KinanBab replied:
I am unsure if this increase readability, I am worried the file will be too big. I do not have a strong preference.

smolkaj commented on test/bmv2_parsing_check.bzl:
```diff
@@ -0,0 +1,67 @@
+    # diff expected with tmp file
+    ctx.actions.write(
+        output = ctx.outputs.executable,
+        content = "{binary} < {input} | python {py_file} {input}".format(
```
I find it a bit confusing that you use abstract names "binary" and "py_file" here when we only every call this with very concrete inputs.

stefanheule commented on src/protobuf/bmv2/table.proto:
```diff
@@ -0,0 +1,107 @@
+// usually there are two pipelines in a p4 program
+// one for ingress and one for egress
+message Pipeline {
+  // name assigned by programmer
```
Start with upper-case letter.

stefanheule commented on src/protobuf/bmv2/table.proto:
```diff
@@ -0,0 +1,107 @@
+  // Full qualified name assigned by programmer
+  // e.g. MyIngressPipeline.MyTable
+  string name = 1;
+  // Auto-generated id
```
Is this really autogenerated? I think this is the P4Runtime table ID, no? It can be generated, but you can also pick it as a programmer (see https://s3-us-west-2.amazonaws.com/p4runtime/docs/master/P4Runtime-Spec.html#sec-id-allocation). It would be good to clarify the comment, here and for other IDs)

smolkaj commented on BUILD:
```diff
@@ -0,0 +1,32 @@
+
+cc_binary(
+    name = "main",
+    srcs = ["src/main.cc"],
```
Please always define targets in a BUILD file located in the same folder as the source files. This is so `bazel build some/folder:all` behaves as expected.

Concretely, this BUILD files should be in the src folder.

stefanheule commented on src/protobuf/bmv2/table.proto:
```diff
@@ -0,0 +1,107 @@
+  // is applied to a packet
+  // can be 'exact', 'lpm', 'ternary'
+  string match_type = 1;  // TODO(babman): make into an enum (same as in Table)
+  // This is essentially a string representation of the target
```
"Name of the match field, e.g. standard_metadata.ingress_port". This can also be user-chosen, using the @name annotation.

stefanheule commented on src/protobuf/bmv2/table.proto:
```diff
@@ -0,0 +1,107 @@
+
+// An action that can be taken when a table is matched on
+message Action {
+  // Full qualified name assigned by programmer
```
Same as above, "Name of the action". It is not fully qualified necessarily, the user can chose a name using @name.

smolkaj commented on test/bmv2_parsing_check.bzl:
```diff
@@ -0,0 +1,67 @@
+        runfiles = ctx.runfiles(files = runfiles),
+    )
+
+subset_diff_test = rule(
```
It seems like this rule does two things: run the binary `main.cc`, and also compute a subset diff of the generated output with another json file. Can we refactor this so that it only does the second thing? I.e., it should take two json files, say `actual` and `expected` and compute the diff.
This will make your rule less complex and more reusable.

Simply use `genrule` to run the binary, no need to write a custom skylark rule for that.

KinanBab replied:
I sort of did this before reading this comment, the rules are now all split into coherent and self-containing pieces.

I have not used genrule, and instead used my own custom rule. I would like to keep it that way for two reasons:
1. The rule that I created autogenerates the output file names based on the input names, which is nice, because the user of the rule does not have to worry about paths and names etc

2. The rule outputs 2 files, and returns them using two different output groups (one json and one protobuf), which are easy to use as dependencies in other rules that require only a single file in a generic way. I do not know how this could be done with genrule, which gives all the output files together using the default provider.

For example, let us say I have defined some rule A responsible for running the binary and producing the json and pb.txt files, and now I want to feed only the json output file to some independent generic rule B defined separately.

if I just define B as
<something>(
   ...
   deps = [A]
)

and A is a plain gen rule, then B will receive both the json and pb.txt files via the default provider, which can cause errors if B is expecting a single file, or may do the wrong thing (e.g. apply B to both files instead of one etc).

The way I have it, the custom rule for A defines two output groups, then I can depend only on one of the two output groups in B's definition. This works even if B expects a single file, and does not know anything about the internals of A. It also does not restrict what providers rules are allowed to return if they are to be used in B.

smolkaj replied:
Sorry, just saw this after suggesting `genrule` above.

To answer your question, you can depend on generated files by name. So if your genrule produces `foo.json` and `bar.proto`, you can simply add `foo.json` and `bar.proto` in the dependency lists of other rules. 

I'll make it more concrete further down in your test rule.

smolkaj commented on test/bmv2_parsing_check.bzl:
```diff
@@ -0,0 +1,67 @@
+            default = "sdiff.py"
+        ),
+        "_binary": attr.label(
+            doc = "src/main.cc binary that uses the protobuf definitions",
```
If we use the binary only for testing, let's move it to the test folder. Let's also give it a more descriptive name.

KinanBab replied:
Yes, this is the plan, but I would like to do it in the next PR (I have an issue open for it).

smolkaj commented on test/output/basic.p4c.json:
```diff
@@ -0,0 +1,558 @@
+{
```
Is this the expected output or the generated output? Could we rename things so it is obvious which one it is?

smolkaj commented on test/sdiff.py:
```diff
@@ -0,0 +1,62 @@
+import sys
+import json
+
+def sdiff(actual, expected, path):
```
Please add a docstring: go/pystyle#function-docs

smolkaj replied:
In particular, what are the arguments? What are their types?

smolkaj commented on test/sdiff.py:
```diff
@@ -0,0 +1,62 @@
+    if expected is None and (
+        actual == 0 or actual == "" or actual == False or \
+        actual is None or actual == [] or actual == {}
+    ):  # equate default values to None
```
Replace expression in parens with `not actual`.

smolkaj replied:
Also, why is this not just `if not expected:`? Why does it matter what `actual` is?

smolkaj commented on test/sdiff.py:
```diff
@@ -0,0 +1,62 @@
+      return (False, path)
+
+    if isinstance(actual, list):
+      if len(actual) != len(expected):
```
`not(isinstance(expected, list)) or len ....`?

smolkaj replied:
So the "subset" refers only to dictionaries but not to lists? May be worth explaining this in the docstring.

smolkaj commented on test/sdiff.py:
```diff
@@ -0,0 +1,62 @@
+        status, rpath = sdiff(actual[i], expected[i], "%s/%d" % (path, i))
+        if not status:
+          return (False, rpath)
+
```
Remove this line? Make is a bit easier to see which level the return is at.

smolkaj commented on test/sdiff.py:
```diff
@@ -0,0 +1,62 @@
+      for i in range(len(actual)):
+        status, rpath = sdiff(actual[i], expected[i], "%s/%d" % (path, i))
+        if not status:
+          return (False, rpath)
```
Maybe replace False with status? Makes it clearer that you're just passing on the result.

smolkaj commented on test/sdiff.py:
```diff
@@ -0,0 +1,62 @@
+
+      return (True, "")
+
+    if isinstance(actual, dict):
```
Insert check that `expected` is also dict.
EDIT: I see this is handled further down.

smolkaj reviewed (changes requested):
Overall this looks great!

I have a lot of small comments, many things are just nits. We have a ton of conventions at Google that we are pretty religious about. It can be annoying initial, but it starts paying off pretty quickly.

As a meta comment, I would encourage you to try to submit things in smaller chunks (see http://go/small-cls), that way we can get the annoying part (conventions) out of the way quicker and focus on deeper issues.

KinanBab committed:
Improve protobuf definitions documentation

KinanBab committed:
Improve documentation and formatting of subset diff test rules and python implementation

KinanBab committed:
Remove test suites from bazel BUILD files

KinanBab committed:
Uncheck json p4c and protobuf output files

KinanBab committed:
Add one more TODO

KinanBab committed:
Change main.cc to dump both json formatted and protobuf text representation of parsed bmv2 programs

KinanBab committed:
Implement golden file testing on protobuf formatted files in addition to subset diff on json files.

* sdiff.py now reads both json files as command line arguments
* implement new bazel rule for exact diffing and updating of golden files
* golden files are stored and cheked in under test/expected
* separate bazel rules for compiling with p4c into bmv2 json, running our test binary on bmv2 json, subset diff on jsons, and exact diff on protobuf dumps

KinanBab committed:
Add macro that summerizes the needed rules to do subset and exact diff testing on p4 programs

smolkaj commented:
Please take a look at my open comments and mark them as resolved once you're done. I'll take another pass afterwards.

stefanheule commented on test/bmv2_parsing_check.bzl:
```diff
@@ -240,3 +242,57 @@ subset_diff_test = rule(
     }
 )
+
+# Macro that defines subset and exact diff rules for a given p4 program
```
Maybe say exactly what the macro does? I.e. what things are called.

stefanheule commented on p4c.bzl:
```diff
@@ -0,0 +1,52 @@
+
+    return [DefaultInfo(files = depset([json_output_file]))]
+
+run_p4c = rule(
```
p4c can output many things, this is specifically only outputing the bmv2 json file, so maybe gen_bmv2_json is a better name?

KinanBab replied:
I actually made this a bit more general now.

stefanheule commented on test/bmv2_parsing_check.bzl:
```diff
@@ -0,0 +1,298 @@
+
+extract_output_group = rule(
+    doc = """Extract the specified output group from a provided target..
+             Always targets to depend on the specified output group only,
```
I'm not sure, but I think there is a typo in this sentence. Doesn't quite make sense right now, I think?

KinanBab committed:
Generalize run_p4c rule and clean up p4c.bzl

Closes #6.

KinanBab committed:
Cleanup main.cc and protobuf definitions and their build rules

KinanBab committed:
Clean up and improve documentation of testing bazel scripts and subset diff python script

KinanBab committed:
Use p4_symbolic as the top level package/namespace in cc and proto files

stefanheule commented on p4c.bzl:
```diff
@@ -12,41 +12,110 @@
+                p4c via -o, relative to the path of the first input file.
+                If not provided, this defaults to an auto generated path, based
+                on the rule name.
+                The output file is not checked into the repo, it is only created
```
What is and isn't checked in seems independent of the build rule that produces it. Lets remove that comment.

smolkaj commented on p4c.bzl:
```diff
@@ -0,0 +1,121 @@
+# This file defines a rule for running p4c and producing bmv2 json files.
+
+def _run_p4c_impl(ctx):
+    # output directory
```
Nit: `# Comment.` (here and below)

smolkaj commented on p4c.bzl:
```diff
@@ -0,0 +1,121 @@
+    fname = "".join([
+        out_dir,
+        "/",
+        ctx.files.srcs[0].basename[:-3],
```
What does this do? Probably worth adding a comment what `basename[:-3]` is.

smolkaj commented on p4c.bzl:
```diff
@@ -0,0 +1,121 @@
+            ctx.attr.std,
+            ctx.attr.target,
+            ctx.attr.arch,
+            output_file.path[:-len(output_file.basename)-1],
```
What does this do?

smolkaj commented on p4c.bzl:
```diff
@@ -0,0 +1,121 @@
+            ctx.attr.arch,
+            output_file.path[:-len(output_file.basename)-1],
+            ctx.attr.p4c_args,
+            " ".join([file.path for file in ctx.files.srcs])
```
What does p4c do when you give it several input files?

KinanBab replied:
oh no sorry, this is my bad, I will show you what I meant in just a minute when I push.

smolkaj commented on p4c.bzl:
```diff
@@ -0,0 +1,121 @@
+            mandatory = False,
+            default = ""
+        ),
+        "output": attr.string(
```
Would we ever want to specify this manually?

KinanBab replied:
no, I do not think we would, but I added it just so we can depend on the output by name if we ever want to

smolkaj commented on src/main.cc:
```diff
@@ -0,0 +1,99 @@
+namespace bmv2_json {
+
+// Read all of stdin up to EOF.
+inline std::string ReadStdin() {
```
Why `inline`?
In general I would abstain from using `inline` (or more generally, adding complexity) unless there is a good reason for it. The compiler can figure this out, but even if it didn't there would be no observable performance difference.

KinanBab replied:
I wanted to avoid the copy of the string. I will remove it.

smolkaj commented on src/protobuf/BUILD:
```diff
@@ -0,0 +1,39 @@
+
+proto_library(
+    name = "p4_parsing_proto",
+    srcs = [
```
The convention at Google is to *always* put the BUILD file in the same folder as the source files. That way there is no ambiguity where to look for the BUILD file, and also no ambiguity for the target name.

smolkaj commented on test/bmv2_parsing_check.bzl:
```diff
@@ -0,0 +1,339 @@
+    basename = "".join([
+        ctx.attr.name,
+        "-parse-bmv2-tmp/",
+        ctx.file.input.basename[:-3]
```
As above, please add a one-liner explaining what this means.

smolkaj commented on test/bmv2_parsing_check.bzl:
```diff
@@ -0,0 +1,339 @@
+def _parse_bmv2_impl(ctx):
+    # come up with a base name in a tmp directory
+    # for the .json and .pb.txt output files
+    basename = "".join([
```
Bazel builds everything in a sandbox by default, so I don't think this is needed?

smolkaj commented on test/bmv2_parsing_check.bzl:
```diff
@@ -0,0 +1,339 @@
+        )
+    ]
+
+parse_bmv2 = rule(
```
Couldn't this just be a macro invoking a genrule?

smolkaj commented on test/bmv2_parsing_check.bzl:
```diff
@@ -0,0 +1,339 @@
+        "actual": attr.label(
+            doc = """The 'Actual' file.
+                     Typically the output of some command.
+                     This is not checked in and is only temporary.""",
```
Is this a suggestion? Technically it's not relevant whether the file is checked in or not.

smolkaj commented on test/bmv2_parsing_check.bzl:
```diff
@@ -0,0 +1,339 @@
+        output_group = "protobuf"
+    )
+
+    parse_bmv2(
```
This is the place where you could invoke a genrule instead. Something like
```
genrule(
  name = parse_name,
  srcs = [":" + p4c_name],
  outs = [proto_name, json_name],
  tools = ["//src:main"],
  cmd = """
    $(location //src:main) "$(OUTS)" < "$(SRCS)"
  """
```

smolkaj commented on test/bmv2_parsing_check.bzl:
```diff
@@ -0,0 +1,339 @@
+        expected = golden_file
+    )
+
+    extract_output_group(
```
If you depend on files by name, these will become obsolete.

smolkaj reviewed (changes requested):
This looks great overall and we should get it merged!

I do think it would be worth reducing the complexity of our bazel rules a bit, though, see my comments. In general, I think it is a good idea to have as little logic as possible in `.bzl` files.
If you would prefer doing that in a separate PR that's okay, I know this PR is huge already and we don't want to slow you down too much.

stefanheule commented:
> This looks great overall and we should get it merged!
> 
> I do think it would be worth reducing the complexity of our bazel rules a bit, though, see my comments. In general, I think it is a good idea to have as little logic as possible in `.bzl` files.
> If you would prefer doing that in a separate PR that's okay, I know this PR is huge already and we don't want to slow you down too much.

Agree with this. I would also add that we are working on open source p4c rules, and once those are available, we should use those instead of running our own.

KinanBab committed:
Move src/protobuf/BUILD to src/protobuf/bmv2/BUILD

KinanBab committed:
Improve p4c.bzl
1. Seperate src from deps.
2. Remove output parameter to simplify implementation and rule.
3. Add better comments

KinanBab committed:
Remove inline

KinanBab committed:
More cleanup and styling

stefanheule reviewed (approved):
Thanks so much for improving this. This is looking great. Lets address the rest using follow-up PRs and track them using issues like we have already done.

smolkaj commented:
+1, nice work

KinanBab committed:
Use genrule instead of custom rules in test/bmv2_parsing_check.bzl
